File: ob1k-concurrent-scala/src/main/scala/com/outbrain/ob1k/concurrent/scalaapi/ComposablePromise.scala
package com.outbrain.ob1k.concurrent.scalaapi

import java.util.concurrent.Executor

import com.outbrain.ob1k.concurrent.eager.{ComposablePromise => JavaComposablePromise}
import com.outbrain.ob1k.concurrent.{ComposableFutures => JavaComposableFutures}

import scala.util.{Failure, Success, Try}

/**
 * A Scala friendly adaptation of the ob1k ComposablePromise API.
 * Inspired by Scala's native Future API.
 *
 * @author slevin
 * @since 26/02/15
 */
object ComposablePromise {

  implicit def toScalaComposablePromise[T](composablePromise: JavaComposablePromise[T]): ComposablePromise[T] = {
    ComposablePromise(composablePromise)
  }

  def apply[T](): ComposablePromise[T] = {
    JavaComposableFutures.newPromise[T]()
  }

  def apply[T](executor: Executor): ComposablePromise[T] = {
    JavaComposableFutures.newPromise[T](executor)
  }

  private def apply[T](composablePromise: JavaComposablePromise[T]): ComposablePromise[T] = {
    new ComposablePromise[T] {
      override val promise = composablePromise
    }
  }
}

trait ComposablePromise[T] {

  protected val promise: JavaComposablePromise[T]

  /** Completes the promise with either an exception or a value.
    *
    * @param tryValue     Either the value or the exception to complete the promise with.
    *
    */
  def complete(tryValue: Try[T]): this.type = {
    tryValue match {
      case Success(v) => promise.set(v)
      case Failure(e) => promise.setException(e)
    }

    this
  }

  /** Completes this promise with the specified future, once that future is completed.
    *
    * @return   This promise
    */
  final def completeWith(other: ComposableFuture[T]): this.type = {
    other consume complete
    this
  }

  /** Completes the promise with either an exception or a value.
    *
    * @param e     Either the value or the exception to complete the promise with.
    */
  def failure(e: Throwable): this.type = complete(Failure(e))

  /** Completes the promise with a value.
    *
    * @param v    The value to complete the promise with.
    *
    */
  def success(v: T): this.type = complete(Success(v))

  /**
   * Returns the future containing the value of this promise.
   */
  def future: ComposableFuture[T] = promise.future()
}

File: ob1k-concurrent/src/main/java/com/outbrain/ob1k/concurrent/eager/ComposablePromise.java
package com.outbrain.ob1k.concurrent.eager;

import com.outbrain.ob1k.concurrent.ComposableFuture;
import com.outbrain.ob1k.concurrent.Try;

/**
 * a promise is the producing side of an eager future.
 * it can be used to set a value or an error inside a future
 * that is returned from the promise.
 *
 * @author aronen
 * Date: 6/10/13
 * Time: 5:25 PM
 */
public interface ComposablePromise<T> {

  /**
   * set a value or error inside the corresponding future.
   * a value(or an error) can only be set once.
   *
   * @param value the value
   */
  void setTry(Try<? extends T> value);

  /**
   * set a value inside the corresponding future.
   * a value(or an error) can only be set once.
   *
   * @param value the value
   */
  void set(T value);

  /**
   * set an error inside the corresponding future.
   * an error(or a value) can only be set once.
   *
   * @param error the error.
   */
  void setException(Throwable error);

  /**
   * returns the corresponding future(always the same one)
   *
   * @return the future.
   */
  ComposableFuture<T> future();
}
